readme.txt
===BP Group Analytics  ===
Contributors: erviveksharma2008
Tags: BuddyPress, group, xprofile fields, plugin, charts
Version: 1.1
Revision Date: September 13, 2019
Requires at least: 3.5.1
Tested up to: 5.2.3
Requires PHP: 7.2
License:  GNU General Public License 3.0 or newer (GPL) http://www.gnu.org/licenses/gpl.html
Author: Vivek Sharma
Author URI: http://ProvisTechnologies.com
Website: http://ProvisTechnologies.com
Stable tag: trunk

Pie charts for xprofile fields.

== Description ==
Due to the presence of large number of groups and members, it becomes difficult to manage and procure relevant information about them. BP group analytics plugin helps solves this problem by providing an overview of the group members by using their xprofile fields.
This plugin helps to understand the relevant information easily through its graphical representation.

== Notes ==

Step 1:- Install the plugin and go to plugin settings in WordPress dashboard.

Step 2:- All xprofile fields are listed in the admin settings page. Select the xprofile fields for which chart is required and save settings

Step 3:- A new analytics tab will be added on groups single page, with pie charts by the profile fields selected in admin settings.


Roadmap.txt - contains ideas proposed and the (approximate) order of implementation

History.txt - contains all the changes since version .1

License.txt - contains the licensing details for this component.

== Feedback ==

Please leave a comment  with any bugs, improvements, or comments at

https://github.com/erviveksharma/bp-group-analytics

== Screenshots ==
Admin settings page
Analytics tab in group single page


history.txt
- Initial Release.
Get chart by xprofile fields.

-1.1-
Update according to the latest version of WordPress and BuddyPress


include/admin.php
<?php

// Exit if accessed directly
if (!defined('ABSPATH'))
    exit;

/**
 * bp_group_analytics_admin()
 *
 * Checks for form submission, saves component settings and outputs admin screen HTML.
 *
 */
function bp_group_analytics_admin() {

    do_action('bp_group_analytics_admin');

    /* If the form has been submitted and the admin referrer checks out, save the settings */
    if (!empty($_POST) && check_admin_referer('bpgroup-analytics-settings-save', 'bpgroup-analytics-settings-nonce_field')) {
        if (isset($_POST['xprofile_selected_fields'])) {
            $post_xprofile_selected_fields = sanitize_option(BP_GROUP_ANALYTICS_OPTIONS_META_TITLE, $_POST['xprofile_selected_fields']);
            update_option('BP_GROUP_ANALYTICS_OPTIONS_META_TITLE', $post_xprofile_selected_fields);
        } else {
            update_option('BP_GROUP_ANALYTICS_OPTIONS_META_TITLE', '');
        }

        $updated = true;
    }

    $xprofile_selected_fields_value = get_option('BP_GROUP_ANALYTICS_OPTIONS_META_TITLE');

    $xprofile_selected_fields = array();
    if(!empty($xprofile_selected_fields_value))
        $xprofile_selected_fields = explode(",",$xprofile_selected_fields_value);

    $profile_groups = BP_XProfile_Group::get( array( 'fetch_fields' => true ) );

    ?>
        <div class="wrap">
            <h2>Buddypress Group Analytics: <?php _e('Settings'); ?></h2>
            <br/>

            <?php
            if (isset($updated))
                echo "<div id='message' class='updated fade'><p>" . __('Settings Updated.', 'bp-group-analytics') . "</p></div>";
            ?>

                <form action="" name="group-analytics-settings-form" id="group-analytics-settings-form" method="post">
                <table class="form-table">
                    <tr valign="top">
                        <th scope="row"><label for="xprofile_selected_fields"><?php _e('Select xprofile fields for charting', 'bp-group-analytics') ?>:</label></th>
                        <td>
                            <select id="xprofile_selected_fields" name="xprofile_selected_fields[]" multiple="multiple" size="10" >
                            <?php
                            if ( !empty( $profile_groups ) ) {
                                foreach ( $profile_groups as $profile_group ) {
                                    if ( !empty( $profile_group->fields ) ) {
                                        foreach ( $profile_group->fields as $field ) {
                                            if(in_array($field->id, $xprofile_selected_fields)){
                                                echo '<option selected="selected" value="'.$field->id.'|'.$field->name. '">' . $field->name . '</option> ';
                                            } else {
                                                echo '<option value="'.$field->id.'|'.$field->name. '">' . $field->name . '</option> ';
                                            }
                                        }
                                    }
                                }
                            }
                            ?>
                            </select>
                        </td>
                    </tr>
                </table>
                <p class="submit">
                    <input type="submit" name="submit" value="<?php _e('Save Settings', 'bp-group-analytics') ?>"/>
                </p>
                    <?php wp_nonce_field('bpgroup-analytics-settings-save', 'bpgroup-analytics-settings-nonce_field'); ?>

                </form>
            </div><!-- .wrap -->
            <?php
        }

/**
 * Finds the url of settings page
 * @global type $wpdb
 * @global type $bp
 * @return string
 *
 */
function bp_group_analytics_find_admin_location() {
    global $wpdb, $bp;
    if (!is_super_admin())
        return false;
    // test for BP1.6+ (truncated to allow testing on beta versions)
    if (version_compare(substr(BP_VERSION, 0, 3), '1.6', '>=')) {
        // BuddyPress 1.6 moves its admin pages elsewhere, so use Settings menu
        $locationMu = 'settings.php';
    } else {
        // versions prior to 1.6 have a BuddyPress top-level menu
        $locationMu = 'bp-general-settings';
    }
    $location = bp_core_do_network_admin() ? $locationMu : 'options-general.php';
    return $location;
}

/**
 *
 * @global type $wpdb
 * @global type $bp
 * @return boolean
 *
 */
function bp_group_analytics_group_add_admin_menu() {
    global $wpdb, $bp;
    /* Add the administration tab under the "Site Admin" tab for site administrators */
    $page = add_submenu_page(
            bp_group_analytics_find_admin_location(), 'Buddypress Group Analytics ' . __('Settings'), '<span class="bp-group-analytics-admin-menu-header">' . __('Buddypress Group Analytics', 'bp-group-analytics') . '</span>', 'manage_options', 'bp-group-analytics-settings', 'bp_group_analytics_admin');
}

add_action(bp_core_admin_hook(), 'bp_group_analytics_group_add_admin_menu', 10);

/**
 * Add settings link on plugin page
 * @param type $links
 * @param type $file
 * @return array
 *
 */
function bp_group_analytics_settings_link($links, $file) {
    $this_plugin = BP_GROUP_ANALYTICS_DIR . '/loader.php';
    if ($file == $this_plugin) {
        return array_merge($links, array(
            'settings' => '<a href="' . add_query_arg(array('page' => 'bp-group-analytics-settings'), bp_group_analytics_find_admin_location()) . '">' . esc_html__('Settings', 'bp-group-analytics') . '</a>',
        ));
    }

    return $links;
}

/*
 * validate and sanitize the posted values by applying filter on options
 * todo put a proper naming method to it.
 */
function bp_group_analytics_xprofile_selected_fields($posted_value, $option){
    $option_value = "";

    //get possible profile fields values
    $profile_groups = BP_XProfile_Group::get( array( 'fetch_fields' => true ) );
    $profile_fields = array();
    if ( !empty( $profile_groups ) ) {
        foreach ( $profile_groups as $profile_group ) {
            if ( !empty( $profile_group->fields ) ) {
                foreach ( $profile_group->fields as $field ) {
                    $profile_fields[] = $field->id.'|'.$field->name;
                }
            }
        }
    }

    //validate posted values if exists in the possible values.
    if(is_array($posted_value)){
        foreach($posted_value as $value) {
            if(!in_array($value, $profile_fields)){
                wp_die("Invalid data, Please check the values you have selected.");
            }
        }
        $option_value = sanitize_text_field(implode(",",$posted_value));
    } else {
        if(!in_array($posted_value, $profile_fields)){
            wp_die("Invalid data, Please check the values you have selected.");
        }
        $option_value = sanitize_text_field($posted_value);
    }

    return $option_value;
}

add_filter( 'sanitize_option_bp_group_analytics_xprofile_selected_fields', 'bp_group_analytics_xprofile_selected_fields', 10, 2 );
/// Add link to settings page
add_filter('plugin_action_links', 'bp_group_analytics_settings_link', 10, 2);
add_filter('network_admin_plugin_action_links', 'bp_group_analytics_settings_link', 10, 2);


buddypress-group-analytics.php
<?php
// Exit if accessed directly
if (!defined('ABSPATH'))
    exit;

/**
 * @author Vivek Sharma
 * @since  1.0
 * @version 1.0
 */
if (class_exists('BP_Group_Extension')) : // Recommended, to prevent problems during upgrade or when Groups are disabled

    class BP_Group_Analytics_Plugin_Extension extends BP_Group_Extension {

        protected $members;

        function __construct() {
            global $bp;
            $this->name =  __('Analytics', 'bp-group-analytics');
            $this->slug = BP_GROUP_ANALYTICS_SLUG;

            /* For internal identification */
            $this->id = 'group_analytics';
            //$this->format_notification_function = 'bp_group_analytics_format_notifications';

            if ($bp->groups->current_group) {
                $this->nav_item_name = $this->name;
                $this->nav_item_position = 101;
            }

            $this->admin_name =  __('Analytics', 'bp-group-analytics');
            $this->admin_slug = BP_GROUP_ANALYTICS_SLUG;

            if(!is_admin())
                $this->members = $this->_get_group_members();
        }

        /*
         * Return group members for the current group or given group ID
         * @param int group ID
         * @return array member IDs
         */
        protected function _get_group_members($group_id = 0){
            global $bp;
            $members = array();

            if (empty($group_id)) {
                $group_id = $bp->groups->current_group->id;
            }

            $has_members_str = array(
                'group_id' => $group_id,
                'per_page' => 0,
                'exclude_admins_mods' => 0,
            );

            if ( bp_group_has_members( $has_members_str ) ) {
                while ( bp_group_members() ) : bp_group_the_member();
                    $members[]= bp_get_group_member_id();
                endwhile;
            }

            return $members;

        }

        /*
         * Get xprofile field data by field ID
         * @param int field ID
         * @return array a array of counts by profile field grouped by values
         */
        protected function _get_xprofile_field_data($field_id){
            global $bp;
            $results = array();
            if(!empty($this->members)){
                foreach($this->members as $member_id){
                    $field_data = xprofile_get_field_data( $field_id , $member_id);
                    $results = $this->_update_item_key($results,$field_data);
                }
            }
            return $results;
        }

        protected function _update_item_key($array = array(), $key = 'N/A'){
            if ($key == "") $key = __('N/A', 'bp-group-analytics');
            if(!empty($array)){
                if (in_array($key, array_keys($array))) {
                    $count = $array[$key][1];
                    $count = $count+1;
                    $array[$key] = array($key,$count);
                } else {
                    $array[$key] = array($key,1);
                }
            } else {
                $array[$key] = array($key,1);
            }
            return $array;
        }

        protected function _display_field_chart($title, $container_id, $data){
            $chartHTML = "";
            $data_variable = $container_id."_data";
            $chartHTML .=  '<h3>'.__($title, 'bp-group-analytics').'</h3><div id="'.$container_id.'" class="chart_style"></div>';
            $chartHTML .= '<script>var '.$data_variable.' = '.wp_json_encode(array_values($data)).' </script>';
            $chartHTML .= $this->_chartScript($container_id, $data_variable);
            return $chartHTML;
        }

        protected function _chartScript($container_id, $data) {

            $chartScript = "<script type='text/javascript'>google.charts.load('current', {'packages':['corechart']});google.charts.setOnLoadCallback(drawChart);
                function drawChart() {
                    var data = new google.visualization.DataTable();
                    data.addColumn('string', 'Field');
                    data.addColumn('number', 'Count');
                    data.addRows(".$data.");
                    var options = {chartArea:{left:20,top:20,bottom:20,right:20,width:'100%',height:'100%'}};
                    var chart = new google.visualization.PieChart(document.getElementById('".$container_id."'));
                    chart.draw(data, options);
                }
            </script>";

            return $chartScript;
        }

        public function display_chart($field_id, $title, $index){
            $field_id = intval($field_id);
            $data = $this->_get_xprofile_field_data($field_id);
            if(empty($data)) return false;
            $contents = $this->_display_field_chart($title,"div_".$field_id,$data);
            return apply_filters("bp_group_analytics_filter_chart_html",$contents, $index);

        }

        function create_screen($group_id = null) { }

        function create_screen_save($group_id = null) { }

        function edit_screen($group_id = null) { }

        function edit_screen_save($group_id = null) {   }

        /**
         * @version 1.0
         * @since version 1.0
         * @author Vivek Sharma
         */
        function display($group_id = null) {
            do_action('bp_group_analytics_display');
            $this->bp_group_analytics_display_from_saved_meta();
        }

        function bp_group_analytics_display_from_saved_meta(){
            $xprofile_selected_fields_value = get_option('BP_GROUP_ANALYTICS_OPTIONS_META_TITLE');
            $xprofile_selected_fields = array();
            if(!empty($xprofile_selected_fields_value)){
                $xprofile_selected_fields = explode(",",$xprofile_selected_fields_value);
                $index = 0;
                foreach($xprofile_selected_fields as $field){
                    $field_data = explode("|",$field);
                    echo $this->display_chart($field_data[0],$field_data[1],$index);
                    $index++;
                }
            } else {
                echo "<p>". __('Please add profile fields from plugin admin settings.', 'bp-group-analytics')."</p>";
            }

        }

    }
    /**
     * @author Vivek Sharma
     * @since 1.0
     * @version 1.0
     */
    function bp_group_analytics_include_files() {
        require ( dirname(__FILE__) . '/include/cssjs.php' );
        require ( dirname(__FILE__) . '/include/admin.php' );
    }

    bp_register_group_extension('BP_Group_Analytics_Plugin_Extension');
endif; // class_exists( 'BP_Group_Extension' )



loader.php
<?php
/*
  Plugin Name: BP Group Analytics
  Plugin URI: wordpress.org/plugins/bp-group-analytics/
  GitHub: https://github.com/erviveksharma/bp-group-analytics

  Description: BP Group Analytics.
  Version: 1.1
  Revision Date: September 13,2019
  Requires at least: WP 3.5.1, BuddyPress 1.6.5
  Tested up to: WP 5.2.3, BuddyPress 4.4.0
  License:  GNU General Public License 3.0 or newer (GPL) http://www.gnu.org/licenses/gpl.html
  Author: Vivek Sharma
  Author URI: http://ProvisTechnologies.com

 */


/* Only load code that needs BuddyPress to run once BP is loaded and initialized. */

//some constants that can be checked when extending this plugin
define('BP_GROUP_ANALYTICS_IS_INSTALLED', 1.1);
define('BP_GROUP_ANALYTICS_VERSION', '1.1');


//allow override of URL slug
if (!defined('BP_GROUP_ANALYTICS_SLUG')) {
    define('BP_GROUP_ANALYTICS_SLUG', 'analytics');
    define('BP_GROUP_ANALYTICS_OPTIONS_META_TITLE', 'bp_group_analytics_xprofile_selected_fields');
}

/**
 * Nifty function to get the name of the directory bp_group_analytics plugin is installed in.
 * @author  Vivek Sharma
 * @version 1
 * @since 1
 */
function bp_group_analytics_dir() {
    if (stristr(__FILE__, '/'))
        $bp_gr_dir = explode('/plugins/', dirname(__FILE__));
    else
        $bp_gr_dir = explode('\\plugins\\', dirname(__FILE__));
    return str_replace('\\', '/', end($bp_gr_dir)); //takes care of MS slashes
}

$bp_gr_dir = bp_group_analytics_dir();

define('BP_GROUP_ANALYTICS_DIR', $bp_gr_dir); //the name of the directory that bp_group_analytics  files are located.

/**
 * @author Vivek Sharma
 * @global type $wpdb
 * @return type
 * @since 1
 * @version 1 added the bp_analytics_load_textdomain() call
 */
function bp_group_analytics_init() {
    global $wpdb;
    if ( is_multisite() && BP_ROOT_BLOG != $wpdb->blogid ) {
        return ;
    }
    if ( ! bp_is_active( 'groups' ) ) {
        return ;
    }
    // Because our loader file uses BP_Component, it requires BP 1.6.5 or greater.
    if (version_compare(BP_VERSION, '1.6.5', '>')) {
        require( dirname(__FILE__) . '/buddypress-group-analytics.php' );
        bp_group_analytics_include_files();
    }
    bp_group_analytics_load_textdomain();
}

add_action('bp_loaded', 'bp_group_analytics_init', 50);

/**
 * @author Vivek Sharma
 * @since 1.0
 * @version 1.0
 */
function bp_group_analytics_load_textdomain() {
    $domain = 'bp-group-analytics';
    // The "plugin_locale" filter is also used in load_plugin_textdomain()
    $locale = apply_filters('plugin_locale', get_locale(), $domain);
    load_textdomain($domain, WP_LANG_DIR . '/bp-group-analytics/' . $domain . '-' . $locale . '.mo');

    if (file_exists(dirname(__FILE__) . '/languages/bp-group-analytics-' . get_locale() . '.mo')) {
        load_plugin_textdomain($domain, false, dirname(plugin_basename(__FILE__)) . '/languages/');
    }
}

# check required wordpress plugins
register_activation_hook( __FILE__, 'bp_group_analytics_import_install' );
function bp_group_analytics_import_install() {
    global $bp;

    # Check whether BP is active and whether Groups component is loaded, and throw error if not
    if(!(function_exists('BuddyPress') || is_a($bp,'BuddyPress')) || !bp_is_active('groups')) {
       echo 'BuddyPress is not installed or the Groups component is not activated. Cannot continue.';
        exit;
    }
}
